kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bytes-simplification.k

```diff
@@ -227,27 +227,35 @@ module BYTES-SIMPLIFICATION [symbolic]
       requires S <Int 0
       [simplification]
 
-    rule [memUpdate-as-concat-inside]:
-      B1:Bytes [ S:Int := B2:Bytes ] =>
-        #range(B1, 0, S) +Bytes B2 +Bytes #range (B1 , S +Int lengthBytes(B2), lengthBytes(B1) -Int (S +Int lengthBytes(B2)))
-      requires 0 <=Int S andBool S +Int lengthBytes(B2) <Int lengthBytes(B1)
-      [simplification(60)]
+    rule [memUpdate-concat-in-left]: (B1 +Bytes B2) [ S := B ] => (B1 [ S := B ]) +Bytes B2
+      requires 0 <=Int S andBool S +Int lengthBytes(B) <=Int lengthBytes(B1)
+      [simplification(40)]
 
-    rule [memUpdate-as-concat-outside-1]:
-      B1:Bytes [ S:Int := B2:Bytes ] => #range(B1, 0, S) +Bytes B2
-      requires 0 <=Int S andBool lengthBytes(B1) <=Int S +Int lengthBytes(B2)
-      [simplification(60)]
+    rule [memUpdate-concat-in-right]: (B1 +Bytes B2) [ S := B ] => B1 +Bytes (B2 [ S -Int lengthBytes(B1) := B ])
+      requires lengthBytes(B1) <=Int S
+      [simplification(40)]
 
     rule [memUpdate-reorder]:
-      B:Bytes [ S1:Int := B1:Bytes] [ S2:Int := B2:Bytes ] => B [ S2 := B2 ] [ S1 := B1 ]
+      B:Bytes [ S1:Int := B1:Bytes ] [ S2:Int := B2:Bytes ] => B [ S2 := B2 ] [ S1 := B1 ]
       requires 0 <=Int S2 andBool S2 <Int S1 andBool S2 +Int lengthBytes(B2) <=Int S1
       [simplification]
 
     rule [memUpdate-subsume]:
-      B:Bytes [ S1:Int := B1:Bytes] [ S2:Int := B2:Bytes ] => B [ S2 := B2 ]
+      B:Bytes [ S1:Int := B1:Bytes ] [ S2:Int := B2:Bytes ] => B [ S2 := B2 ]
       requires 0 <=Int S2 andBool S2 <=Int S1 andBool S1 +Int lengthBytes(B1) <=Int S2 +Int lengthBytes(B2)
       [simplification]
 
+    rule [memUpdate-as-concat-inside]:
+      B1:Bytes [ S:Int := B2:Bytes ] =>
+        #range(B1, 0, S) +Bytes B2 +Bytes #range(B1 , S +Int lengthBytes(B2), lengthBytes(B1) -Int (S +Int lengthBytes(B2)))
+      requires 0 <=Int S andBool S +Int lengthBytes(B2) <Int lengthBytes(B1)
+      [simplification(60)]
+
+    rule [memUpdate-as-concat-outside]:
+      B1:Bytes [ S:Int := B2:Bytes ] => #range(B1, 0, S) +Bytes B2
+      requires 0 <=Int S andBool lengthBytes(B1) <=Int S +Int lengthBytes(B2)
+      [simplification(60)]
```

Comment:
I feel like these can be combined into one lemma (so no branching on `S +Int lengthBytes(B2) <Int lengthBytes(B1)` basically):

```
rule B1 [ S := B2 ] => #range(B1, 0, S) +Bytes B2 +Bytes #range(B1, S +Int lengthBytes(B2), lengthBytes(B1) -Int (S +Int lengthBytes(B2))) requires 0 <=Int S andBool S <=Int lengthBytes(B1) [simplification(60)]
```

I think that should handle both cases? The last `#range(...)` expression should go to empty if it can determine that the third argument is negative.

Reply:
Yes, it will. I can't remember right now if there was a reason for splitting, that is, requiring that we know the branch before simplifying. I certainly prefer `+Bytes` over memory update, so let's try this and see what happens.

Reply:
I think we might not need `S <=Int lengthBytes(B1)` either, because if it's greater `#range(B1, 0, S)` will pad with zeros (as intended) and the last `#range(...)` expression goes to `.Bytes`?

Reply:
Yeah, like I said, my only concern in omitting that condition is that we could get a potentially huge 0-bytes expression there, which could slow things down.

Reply:
This is happening right now with the lemmas as they are currently, and there are no issues, we're ok with reasoning about `+Bytes` and large constants.